### PR TITLE
refactor: Clean up nScriptCheckThreads

### PR DESCRIPTION
**Pull request by jnewbery**
The meaning of this value is confusing. Refactor it and add comments.

### CODE REVIEW

**Comment by practicalswift** [comment]
ACK 6775a18cab043af0a9ea69a364804d6d75faf939 -- agree regarding the confusion

**Comment by laanwj** [comment]
I think it would be much more straightforward if 0 was the disallowed value, not 1.

**Comment by jnewbery** [comment]
@laanwj 
> I think it would be much more straightforward if...

Turns out that a global int is not needed at all. We only need the bool-ness of whether any script checking threads have been spawned. I have a branch here: https://github.com/jnewbery/bitcoin/tree/2019-11-nScriptCheckThreads2 that changes the int for a bool, improves commenting everywhere and makes the code clearer.

I fear it'd be NACKed by the "don't change code for purely cosmetic reasons" crowd, but if you prefer it, let me know.

**Comment by promag** [comment]
@jnewbery FWIW the other branch LGTM.

**Comment by laanwj** [comment]
Refactor LGTM, we've both been in the group of people that were confused by this code at some point.

> Turns out that a global int is not needed at all. We only need the bool-ness of whether any script checking threads have been spawned.

Wonder if we can take this a step further and not have any global at all, passing the flag into the ChainState constructor instead.

**Comment by jnewbery** [comment]
Thanks for the feedback everyone. I've changed this PR so it now refactors `nScriptCheckThreads` and adds clarifying comments.

> Wonder if we can take this a step further and not have any global at all, passing the flag into the ChainState constructor instead.

Yes, it'd be great to remove more globals and make validation initialization take these as arguments, but I think that's too far to go in this PR.

**Comment by jnewbery** [comment]
Tested that number of script threads started is the same before and after this PR:

(`bitcoind-pre17342` is a bitcoind compiled from commit bc3fcf3c0d1d4c83da0fab14d40f92ffca35789e in master and `bitcoind-post17342` is a bitcoind compiled from this branch)

```
for par in {-4..17}; do echo par=$par; ./bitcoind-pre17342 -par=$par && sleep 2 && bitcoin-cli stop && grep "Script verification" .bitcoin/regtest/debug.log && rm .bitcoin/regtest/debug.log && sleep 1; ./bitcoind-post17342 -par=$par && sleep 2 && bitcoin-cli stop && grep "Script verification" .bitcoin/regtest/debug.log && rm .bitcoin/regtest/debug.log && sleep 1;done
par=-4
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:30:57Z Script verification uses 0 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:00Z Script verification uses 0 additional threads
par=-3
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:03Z Script verification uses 0 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:06Z Script verification uses 0 additional threads
par=-2
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:10Z Script verification uses 1 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:13Z Script verification uses 1 additional threads
par=-1
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:16Z Script verification uses 2 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:19Z Script verification uses 2 additional threads
par=0
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:22Z Script verification uses 3 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:26Z Script verification uses 3 additional threads
par=1
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:29Z Script verification uses 0 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:32Z Script verification uses 0 additional threads
par=2
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:35Z Script verification uses 1 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:31:38Z Script verification uses 1 additional threads
[...]
ar=15
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:32:58Z Script verification uses 14 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:33:01Z Script verification uses 14 additional threads
par=16
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:33:04Z Script verification uses 15 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:33:07Z Script verification uses 15 additional threads
par=17
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:33:11Z Script verification uses 15 additional threads
Bitcoin Core starting
Bitcoin Core stopping
2019-11-04T19:33:14Z Script verification uses 15 additional threads

```

File: src/validation.cpp [hunk 1]
```diff
@@ -107,7 +107,7 @@ CBlockIndex *pindexBestHeader = nullptr;
 Mutex g_best_block_mutex;
 std::condition_variable g_best_block_cv;
 uint256 g_best_block;
-int nScriptCheckThreads = 0;
+bool g_parallel_script_checks {false};
```
Comment (promag) [inline src/validation.cpp]: 1856073d6e8c774cc0a606275f264133884ef7bd

nit, is this a new format? The space makes sense when the initialization comes after an annotation but in this case we don't add the space.

File: src/test/setup_common.cpp [hunk 2]
```diff
@@ -102,9 +102,12 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
         throw std::runtime_error(strprintf("ActivateBestChain failed. (%s)", FormatStateMessage(state)));
     }
 
-    nScriptCheckThreads = 3;
-    for (int i = 0; i < nScriptCheckThreads - 1; i++)
+    // Start script-checking threads. Set g_parallel_script_checks to true so they are used.
+    constexpr int script_check_threads = 2;
+    for (int i = 0; i < script_check_threads; i++) {
```
Comment (promag) [inline src/test/setup_common.cpp]: nit, since you are touching this, `++i`?

File: src/test/checkqueue_tests.cpp [hunk 3]
```diff
@@ -149,7 +148,7 @@ static void Correct_Queue_range(std::vector<size_t> range)
 {
     auto small_queue = MakeUnique<Correct_Queue>(QUEUE_BATCH_SIZE);
     boost::thread_group tg;
-    for (auto x = 0; x < nScriptCheckThreads; ++x) {
```
Comment (promag) [inline src/test/checkqueue_tests.cpp]: 6e6cb418accf9286148cff9f1966fca27d38e0d7

Locally it builds if `#include <validation.h>` is removed.

**Review by promag — COMMENTED** [review]
Tested ACK 1856073d6e8c774cc0a606275f264133884ef7bd.

This is not a breaking change since the configured `-par` is clamped so existing configurations will work. Also verified that `bitcoin-qt` does work if configured with 16 before this change - the setting `nThreadsScriptVerif` remains 16 but the GUI also clamps to 15.

File: src/test/setup_common.cpp (hunk 2 above)
Comment (laanwj) [reply src/test/setup_common.cpp]: come on, this doesn't matter for integers

**Comment by laanwj** [comment]
code review ACK 1856073d6e8c774cc0a606275f264133884ef7bd

File: src/validation.cpp (hunk 1 above)
Comment (jnewbery) [reply src/validation.cpp]: fixed

File: src/test/setup_common.cpp (hunk 2 above)
Comment (jnewbery) [reply src/test/setup_common.cpp]: fixed

File: src/test/checkqueue_tests.cpp (hunk 3 above)
Comment (jnewbery) [reply src/test/checkqueue_tests.cpp]: fixed

**Comment by jnewbery** [comment]
I wanted to remove the unused `validation.h` include, so I picked up @promag's other changes along the way.

Thanks for the review!

**Comment by sipa** [comment]
ACK 5506ecfe7a65d5705616bc048f2f1735b89993fb

**Comment by promag** [comment]
ACK 5506ecfe7a65d5705616bc048f2f1735b89993fb, only change was addressing my nits.

**Comment by laanwj** [comment]
Code review ACK 5506ecfe7a65d5705616bc048f2f1735b89993fb

**Comment by MarcoFalke** [comment]
ACK 5506ecfe7a65d5705616bc048f2f1735b89993fb 🥐

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 5506ecfe7a65d5705616bc048f2f1735b89993fb 🥐
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiO9QwAz6nYOBxbVWtpsV2BAXLethFVuhGT9eD0joGWjF3ZFJpwXxeEckKYM3Yd
8HyfPK8ZQ0vq2I7rQUwUSqPoacnSvEIhFKx34CAMtc+TLnI0g4HB4vXwDw1Ifjcw
3ljujzu3lW255jARdKo+1VTKSL7CpDKlKn1vsJFviiNi0DLRJ2UFjSrYxvYVrGgb
8s2hIIwjFF41iT0FoSfati5OE1A6wYPJlUAPX6I5MDAwaFiDfMPeTlDOEaXHrG7V
EsfICRMKLF4ahH+HN9x1Ce0jIeyB1jGsU9EAQZT1NEDj3kwsuvmVS4RBfhpZauKE
KLkbLRm32L5fKi5N/xmqyODxW4XKDzAkvDnncpOuS7xHMqbBF4OfI9wUY2jjarhb
vWCBgGNe+nzc4FMU58+Nd4Q/yjXWUrFRA2fUKoZT2MZmgjz1SfMdzWzJMqK9Q51c
ZL0l/SKJnMpy6OqjwgYExmbgP+wvhEqL9ywBxc97wqSToizF6Ex7lvNpz/9+3u/K
F9wuppNR
=bJqw
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6804ad6c56d3cf372758297498bf984f409018ae306e96929df7d927368d1bd1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086804ad6c56d3cf372758297498bf984f409018ae306e96929df7d927368d1bd1f0108791f3eb82d9632a495ca8a9c73bffba08fff01071181db2f5432fbfed1be33f200de05008f1045dc432fef008747b932932a79de10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010737712ef4fc9e7ee33df3efcc4693c3008f0203fb68907091c7d0f959b5cda438e20ba812d031cae99d6513312510d4e9b02e508f1045dc432fdf00819f1121481f076e30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a8dc6ac8a1e516d11c1710fbaa04d0cc08f1045dc432fef00856a8625a97ce6c3a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010d642d53e13b393052501af0cce1cdc2608f1045dc432fdf0089a229aac8464e31b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/test/setup_common.cpp [hunk 4]
```diff
@@ -102,9 +102,12 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
         throw std::runtime_error(strprintf("ActivateBestChain failed. (%s)", FormatStateMessage(state)));
     }
 
-    nScriptCheckThreads = 3;
-    for (int i = 0; i < nScriptCheckThreads - 1; i++)
+    // Start script-checking threads. Set g_parallel_script_checks to true so they are used.
+    constexpr int script_check_threads = 2;
```
Comment (MarcoFalke) [inline src/test/setup_common.cpp]: style-nit: compile time constants are generally ALL_UPPER_CASE

**Pull request merged**